WordPress/gutenberg
# Match Container's User and Group to Host

## What?

This pull request changes some of the permissions for `wp-env` containers in order to ensure parity with the host operating system.

## Why?

By using the host's user and group we are able to make sure the host and containers are all acting as the same owner for files and folders. This ensures files created by any of them are readable and writeable by the rest. For instance, `wp rewrite structure '/%postname%/' --hard` in the `cli` container will create a `.htaccess` file that the `wordpress` container is able to read.

Closes https://github.com/WordPress/gutenberg/issues/22515, https://github.com/WordPress/gutenberg/issues/45592, https://github.com/WordPress/gutenberg/issues/49373, and https://github.com/WordPress/gutenberg/issues/28201

## How?

We add the current host user to the `wordpress` and `tests-wordpress` containers using the same `uid` and `gid` from the host. We also use the `APACHE_RUN_USER` and `APACHE_RUN_GROUP` environment variables so that the web server will run as the host user too. Lastly, when using `wp-env run` we use the host user that we created on the container.

Thanks to https://github.com/WordPress/gutenberg/pull/42826 we can be sure that the _entire_ contents of our WordPress installation is mounted. This is good because it means we don't have to change the permissions of any files or folders to get this to work. The end result is that on operating systems where the distinction matters, the owner of these files is the same everywhere.

## Testing Instructions

1. Since CI is using Linux, successfully passing the tests that use `wp-env` is a good indicator that nothing has broken.
2. Run `npm run env start` locally and confirm that it builds the container. If you already have one, make sure to use `npm run env start --update` so that the new Dockerfile is used.
3. Run `npm run env run wordpress ls -la` and confirm that the files are owned by your current user. On macOS the ownership seemed oddly flaky at times, but, it doesn't matter because of the way the mounting works. I'm not sure if it does this on Linux, but, someone can test :smile:
4. Run `npm run env run cli ls -la` and confirm that the files are still owned by your current user. The name will be missing, but, the raw `uid` is a good indicator of the difference made here.
5. Run `npm run env run wordpress ls -la wp-content/plugins/gutenberg` and note that your current user owns all of the files and folders here.
6. Log into `wp-admin` in the container and install a plugin via the UI. It should work.
7. Create a `wp-cli.yml` file in `/var/www/html` with the content:
```yml
apache_modules:
  - mod_rewrite
```
8. Run `npm run env run cli wp rewrite structure '/%postname%/' --hard`. Use `npm run env run cli ls -la` to confirm that a `.htaccess` file was created as expected.
9. Visit a page in the container and verify that pretty permalink are enabled and working.
10. Upload some media and verify that it works as expected.

Given the opportunity for this PR to go wrong, I would suggest trying anything else that you can think of!
